File: lab-2.4/c++/ExpressionCalculator.cpp
#include "ExpressionCalculator.h"

ExpressionCalculator::ExpressionCalculator(double a, double c, double d)
    : a(a), c(c), d(d) {
}

void ExpressionCalculator::setA(double value) { a = value; }
void ExpressionCalculator::setC(double value) { c = value; }
void ExpressionCalculator::setD(double value) { d = value; }

double ExpressionCalculator::getA() const { return a; }
double ExpressionCalculator::getC() const { return c; }
double ExpressionCalculator::getD() const { return d; }

double ExpressionCalculator::calculateSquareRoot() const {
    double expression = 23 * a;
    if (expression < 0) {
        throw std::invalid_argument("Square root of negative number");
    }
    return sqrt(expression);
}

double ExpressionCalculator::calculateExpression() const {
    try {
        double denominator = (a / 4.0) - 1.0;
        if (fabs(denominator) < 1e-10) {
            throw std::runtime_error("Division by zero");
        }

        double sqrtValue = calculateSquareRoot();
        return (2 * c - d + sqrtValue) / denominator;
    }
    catch (const std::invalid_argument& e) {
        throw; // Re-throw the exception
    }
}

File: lab-2.4/c#/ExpressionCalculator.cs
using System;

namespace ExpressionCalculatorLibrary
{
    public class ExpressionCalculator
    {
        private double a;
        private double c;
        private double d;

        public ExpressionCalculator(double a, double c, double d)
        {
            this.a = a;
            this.c = c;
            this.d = d;
        }

        public double A
        {
            get => a;
            set => a = value;
        }

        public double C
        {
            get => c;
            set => c = value;
        }

        public double D
        {
            get => d;
            set => d = value;
        }

        private double CalculateSquareRoot()
        {
            double expression = 23 * a;
            if (expression < 0)
            {
                throw new ArgumentException("Square root of negative number");
            }
            return Math.Sqrt(expression);
        }

        public double CalculateExpression()
        {
            try
            {
                double denominator = (a / 4.0) - 1.0;
                if (Math.Abs(denominator) < 1e-10)
                {
                    throw new DivideByZeroException("Division by zero");
                }

                double sqrtValue = CalculateSquareRoot();
                return (2 * c - d + sqrtValue) / denominator;
            }
            catch (ArgumentException ex)
            {
                throw new InvalidOperationException("Invalid expression parameters", ex);
            }
        }
    }
}